### PR TITLE
Core 1016 authentication project (work-in-progress)

### PR DESCRIPTION
**Pull request by mthambipillai**

### CODE REVIEW

File: src/java/com/unifina/security/ApiKeyAuthorizationHeader.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+public class ApiKeyAuthorizationHeader extends AuthorizationHeader{
+	private String key;
+
+	public ApiKeyAuthorizationHeader(String key){
+		this.key = key;
+	}
+
+	@Override
+	public String getHeaderName(){
+		return "Token";
```
Comment (hpihkala) [inline src/java/com/unifina/security/ApiKeyAuthorizationHeader.java]: Don't know if this matters, but in our docs it's lower case: `token`

File: src/java/com/unifina/security/SessionToken.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import org.apache.commons.lang.RandomStringUtils;
+import org.joda.time.DateTime;
+
+import java.io.*;
+import java.util.Base64;
+
+public class SessionToken{
+	private String token;
+	private SecUser user;
+	private DateTime expiration;
```
Comment (hpihkala) [inline src/java/com/unifina/security/SessionToken.java]: We tend to use plain old `java.util.Date` objects as they automatically play nice with Grails - might be worth considering here, if no complex datetime manipulation is needed

File: src/java/com/unifina/security/SessionTokenAuthorizationHeader.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+	}
+
+	@Override
+	public String getHeaderName() {
+		return "Bearer";
+	}
+
+	@Override
+	public AuthenticationResult computeAuthenticationResult() {
+		return null;
```
Comment (hpihkala) [inline src/java/com/unifina/security/SessionTokenAuthorizationHeader.java]: Still WIP, but tip: you can access Spring beans statically via
`SessionService sessionService = Holders.getApplicationContext().getBean(SessionService.class);`

File: test/integration/com/unifina/service/SessionServiceIntegrationSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+	void setup() {
+		service = Holders.getApplicationContext().getBean(SessionService)
+
+		List<String> hosts = MapTraversal.getList(Holders.getConfig(), "streamr.redis.hosts");
+		String password = MapTraversal.getString(Holders.getConfig(), "streamr.redis.password");
+
+		Assert.notNull(hosts, "streamr.redis.hosts is null!")
+		Assert.notEmpty(hosts, "streamr.redis.hosts is empty!")
+		Assert.notNull(password, "streamr.redis.password is null!")
```
Comment (hpihkala) [inline test/integration/com/unifina/service/SessionServiceIntegrationSpec.groovy]: ok to be null

File: test/integration/com/unifina/service/SessionServiceIntegrationSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+			username: "username",
+			password: "password",
+			name: "name",
+			email: "email@email.com",
+			timezone: "timezone"
+		).save(failOnError: true, validate: false)
+		SessionToken token = service.generateToken(user)
+		DateTime expectedExpiration = new DateTime().plusHours(SessionService.TTL_HOURS)
+		then:
+		token.getExpiration().getMillis() - expectedExpiration.getMillis() < 500
```
Comment (hpihkala) [inline test/integration/com/unifina/service/SessionServiceIntegrationSpec.groovy]: These two lines testing token properties are not a concern of the integration test, they sound like they belong to a unit test for `SessionToken`.

File: test/unit/com/unifina/controller/api/LoginApiControllerSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+		request.requestURI = "/api/v1/login/challenge"
+		request.method = "POST"
+		withFilters(action: "challenge") {
+			controller.challenge()
+		}
+
+		then:
+		response.status == 200
+		response.json == [
+				id       : null,
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/LoginApiControllerSpec.groovy]: Why should id be null and not 123?

File: test/unit/com/unifina/controller/api/LoginApiControllerSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+		withFilters(action: "response") {
+			controller.response()
+		}
+
+		then:
+		1 * challengeService.verifyChallengeResponse(challenge.id, challenge.challenge, signature, address) >> true
+		1 * ethereumIntegrationKeyService.getOrCreateFromEthereumAddress(address) >> user
+		1 * sessionService.generateToken(user) >> sk
+		response.status == 200
+		response.json == [
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/LoginApiControllerSpec.groovy]: If you do the `toMap()` thing, this simplifies to `response.json == sk.toMap()` (also elsewhere in this file)

File: test/unit/com/unifina/filters/UnifinaCoreAPIFiltersSpec.groovy [hunk 8]
```diff
@@ -57,4 +61,32 @@ class UnifinaCoreAPIFiltersSpec extends Specification {
 		then:
 		response.status == 200
 	}
+
+	void "authentication passes when session token provided"() {
+		given:
+		defineBeans {
+			sessionService(MethodInvokingFactoryBean) {
```
Comment (hpihkala) [inline test/unit/com/unifina/filters/UnifinaCoreAPIFiltersSpec.groovy]: Not 100% sure what's happening here, but if you want to replace beans with whatever mocks or stubs, you could make the test class extend `BeanMockingSpecification` instead of `Specification` and use some of the utils in there

File: grails-app/services/com/unifina/service/EthereumIntegrationKeyService.groovy [hunk 9]
```diff
@@ -109,15 +112,20 @@ class EthereumIntegrationKeyService {
...
-				timezone: "UTC"
-			).save(failOnError: true, flush: true, validate: false)
+		if (key == null) {
+			Calendar now = Calendar.getInstance()
+			TimeZone timeZone = now.getTimeZone()
+			SecUser user = userService.createUser([
+				username       : address,
+				password       : RandomStringUtils.random(32),
+				name           : address,
+				timezone       : timeZone.getDisplayName(),
```
Comment (hpihkala) [inline grails-app/services/com/unifina/service/EthereumIntegrationKeyService.groovy]: This will just return server timezone. We can't know the user timezone: the previous version where you gave UTC here was ok. We will soon get rid of server-side timestamp rendering anyway (#511), so no big deal!

File: grails-app/services/com/unifina/service/SessionService.groovy [hunk 10]
```diff
@@ -1,49 +1,51 @@
...
+	private RedisClient redisClient
+	private RedisAsyncConnection<String, String> connection
 
 	SessionToken generateToken(SecUser user) {
 		SessionToken sk = new SessionToken(TOKEN_LENGTH, user, TTL_HOURS)
-		RedisConnection<String, String> connection = getRedisClient().connect()
-		connection.setex(sk.getToken(), TTL_HOURS * 3600, sk.getUser().id.toString())
+		getConnection().setex(sk.getToken(), TTL_HOURS * 3600, sk.getUser().id.toString())
 		return sk
+		redisClient.connectAsync().expireat()
```
Comment (hpihkala) [inline grails-app/services/com/unifina/service/SessionService.groovy]: unreachable code + why new connection? Maybe this line is some random leftover?

File: grails-app/services/com/unifina/service/SessionService.groovy [hunk 11]
```diff
@@ -1,49 +1,51 @@
...
 
-	RedisClient redisClient
+	private RedisClient redisClient
+	private RedisAsyncConnection<String, String> connection
 
 	SessionToken generateToken(SecUser user) {
 		SessionToken sk = new SessionToken(TOKEN_LENGTH, user, TTL_HOURS)
-		RedisConnection<String, String> connection = getRedisClient().connect()
-		connection.setex(sk.getToken(), TTL_HOURS * 3600, sk.getUser().id.toString())
+		getConnection().setex(sk.getToken(), TTL_HOURS * 3600, sk.getUser().id.toString())
```
Comment (hpihkala) [inline grails-app/services/com/unifina/service/SessionService.groovy]: Should we wait (block) for the request to complete? Otherwise the following order of events is possible:
- Token is created and Redis request is fired with the non-blocking async client
- Token is returned to user
- User attempts to use token, but it is not yet written to redis!

File: grails-app/controllers/com/unifina/controller/api/IntegrationKeyApiController.groovy [hunk 12]
```diff
@@ -60,9 +60,15 @@ class IntegrationKeyApiController {
 
 	@StreamrApi
 	def delete(String id) {
-		ethereumIntegrationKeyService.delete(id, apiUser())
-		response.status = 204
-		render ""
+		boolean statusOk = ethereumIntegrationKeyService.delete(id, apiUser())
+		if(statusOk) {
```
Comment (hpihkala) [inline grails-app/controllers/com/unifina/controller/api/IntegrationKeyApiController.groovy]: code style

File: src/java/com/unifina/security/TokenAuthenticator.java [hunk 13]
```diff
@@ -2,51 +2,79 @@
...
-
-		if (key == null) {
+		if (header == null) {
 			return new AuthenticationResult(true, false);
 		}
-
-		Key keyObject = (Key) InvokerHelper.invokeMethod(Key.class, "get", key);
-		if (keyObject != null) {
-			return new AuthenticationResult(keyObject);
+		if (header[0].equals("Token")) {
```
Comment (hpihkala) [inline src/java/com/unifina/security/TokenAuthenticator.java]: Make case-insensitive `header[0].equalsIgnoreCase(...)`

File: src/java/com/unifina/security/TokenAuthenticator.java [hunk 14]
```diff
@@ -2,51 +2,79 @@
...
+			AuthenticationResult res = getResultFromApiKey(header[1]);
+			res.setAuthorizationHeaderName("Token");
+			return res;
+		} else if (header[0].equals("Bearer")) {
```
Comment (hpihkala) [inline src/java/com/unifina/security/TokenAuthenticator.java]: Make case-insensitive `header[0].equalsIgnoreCase(...)`

File: test/integration/com/unifina/service/TokenAuthenticatorIntegrationSpec.groovy [hunk 15]
```diff
@@ -1,14 +1,13 @@
...
-import grails.test.mixin.Mock
-import spock.lang.Specification
+import com.unifina.security.TokenAuthenticator
+import grails.test.spock.IntegrationSpec
 
 import javax.servlet.http.HttpServletRequest
 
-@Mock([Key])
-class TokenAuthenticatorSpec extends Specification {
+class TokenAuthenticatorIntegrationSpec extends IntegrationSpec {
```
Comment (hpihkala) [inline test/integration/com/unifina/service/TokenAuthenticatorIntegrationSpec.groovy]: I would: 
- leave this as a unit test
- mock `SessionService` (for examples, see subclasses of `BeanMockingSpecification`)
- add a test for the `Bearer` case

File: test/unit/com/unifina/controller/api/CanvasApiControllerSpec.groovy [hunk 16]
```diff
@@ -40,6 +41,7 @@ class CanvasApiControllerSpec extends Specification {
 	static doWithSpring = {
 		springSecurityService(SpringSecurityService)
 		userService(UserService)
+		sessionService(SessionService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/CanvasApiControllerSpec.groovy]: Rather use a mock than the real service? (Yeah, this old test uses real versions of `SpringSecurityService` and `UserService` while it shouldn't - no need to fix that retroactively, but better not keep expanding the set of collaborators)

File: test/unit/com/unifina/controller/api/CategoryApiControllerSpec.groovy [hunk 17]
```diff
@@ -11,6 +13,14 @@ import spock.lang.Specification
 @Mock([Category, UnifinaCoreAPIFilters, SpringSecurityService])
 class CategoryApiControllerSpec extends Specification {
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/CategoryApiControllerSpec.groovy]: Rather mocks than real services?

File: test/unit/com/unifina/controller/api/DashboardApiControllerSpec.groovy [hunk 18]
```diff
@@ -33,6 +34,7 @@ class DashboardApiControllerSpec extends Specification {
 	static doWithSpring = {
 		springSecurityService(SpringSecurityService)
 		userService(UserService)
+		sessionService(SessionService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/DashboardApiControllerSpec.groovy]: mock

File: test/unit/com/unifina/controller/api/DashboardItemApiControllerSpec.groovy [hunk 19]
```diff
@@ -24,6 +25,14 @@ class DashboardItemApiControllerSpec extends Specification {
 	SecUser me
 	List<Dashboard> dashboards
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/DashboardItemApiControllerSpec.groovy]: mocks

File: test/unit/com/unifina/controller/api/IntegrationKeyApiControllerSpec.groovy [hunk 20]
```diff
@@ -1,23 +1,32 @@
...
 @TestFor(IntegrationKeyApiController)
-@Mock([UnifinaCoreAPIFilters, Challenge, Key, SecUser, IntegrationKey])
+@Mock([UnifinaCoreAPIFilters, Key, SecUser, IntegrationKey])
 class IntegrationKeyApiControllerSpec extends Specification {
 	EthereumIntegrationKeyService ethereumIntegrationKeyService
 	SecUser me
 	SecUser someoneElse
 
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/IntegrationKeyApiControllerSpec.groovy]: mocks

File: test/unit/com/unifina/controller/api/KeyApiControllerSpec.groovy [hunk 21]
```diff
@@ -23,6 +24,14 @@ class KeyApiControllerSpec extends Specification {
 
 	SecUser loggedInUser
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/KeyApiControllerSpec.groovy]: mocks

File: test/unit/com/unifina/controller/api/LoginApiControllerSpec.groovy [hunk 22]
```diff
@@ -1,29 +1,43 @@
...
 	ChallengeService challengeService
 	SessionService sessionService
 	EthereumIntegrationKeyService ethereumIntegrationKeyService
 	UserService userService
+	DateFormat df
+
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/LoginApiControllerSpec.groovy]: mockety mock

File: test/unit/com/unifina/controller/api/ModuleApiControllerSpec.groovy [hunk 23]
```diff
@@ -22,6 +24,14 @@ class ModuleApiControllerSpec extends Specification {
 	ModulePackage modulePackage
 	Module module
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/ModuleApiControllerSpec.groovy]: mockadoodle

File: test/unit/com/unifina/controller/api/PermissionApiControllerSpec.groovy [hunk 24]
```diff
@@ -32,6 +34,14 @@ class PermissionApiControllerSpec extends Specification {
 	Permission canvasPermission, streamPermission, canvasAnonPermission
 	List<Permission> ownerPermissions
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/PermissionApiControllerSpec.groovy]: Mocky McMockface

File: test/unit/com/unifina/controller/api/RelatedProductsControllerSpec.groovy [hunk 25]
```diff
@@ -41,6 +44,14 @@ class RelatedProductsControllerSpec extends Specification {
 	SecUser u1, u2, apiUser
 	Category cat1, cat2
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/RelatedProductsControllerSpec.groovy]: let's solve this

File: test/unit/com/unifina/controller/api/RemoveUsersProductsControllerSpec.groovy [hunk 26]
```diff
@@ -17,6 +19,14 @@ class RemoveUsersProductsControllerSpec extends Specification {
 	ProductService productService
 	SecUser me
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/RemoveUsersProductsControllerSpec.groovy]: repeat

File: test/unit/com/unifina/controller/api/StreamApiControllerSpec.groovy [hunk 27]
```diff
@@ -32,6 +32,14 @@ class StreamApiControllerSpec extends Specification {
 	def streamThreeId
 	def streamFourId
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/StreamApiControllerSpec.groovy]: repeat

File: test/unit/com/unifina/controller/api/UserApiControllerSpec.groovy [hunk 28]
```diff
@@ -16,6 +17,14 @@ class UserApiControllerSpec extends Specification {
 
 	SecUser me
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/controller/api/UserApiControllerSpec.groovy]: repeat

File: test/unit/com/unifina/feed/PollingMessageSourceSpec.groovy [hunk 29]
```diff
@@ -1,12 +1,23 @@
...
 import spock.util.concurrent.PollingConditions
 
 class PollingMessageSourceSpec extends Specification {
 
 	List<PollingMessageSource<String,String>> sources = []
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/feed/PollingMessageSourceSpec.groovy]: repeat

File: test/unit/com/unifina/service/ChallengeServiceSpec.groovy [hunk 30]
```diff
@@ -1,70 +1,29 @@
...
 import grails.test.mixin.TestFor
+import org.joda.time.DateTime
 import spock.lang.Specification
 
 /**
  * See the API for {@link grails.test.mixin.services.ServiceUnitTestMixin} for usage instructions
  */
 @TestFor(ChallengeService)
-@Mock(Challenge)
 class ChallengeServiceSpec extends Specification {
```
Comment (hpihkala) [inline test/unit/com/unifina/service/ChallengeServiceSpec.groovy]: How can this work as a unit test? Are you sure it's not connecting to redis? If it is, should be an integration test

File: test/unit/com/unifina/service/EthereumIntegrationKeyServiceSpec.groovy [hunk 31]
```diff
@@ -269,4 +265,17 @@ class EthereumIntegrationKeyServiceSpec extends Specification {
...
+		when:
+		String address = "0x26e1ae3f5efe8a01eca8c2e9d3c32702cf4bead6"
+		SecUser user = new SecUser(username: address).save(failOnError: true, validate: false)
+		IntegrationKey integrationKey = new IntegrationKey(
+			user: user,
+			idInService: address,
+			service: IntegrationKey.Service.ETHEREUM_ID
+		).save(failOnError: true, validate: false)
+		then:
+		service.delete(integrationKey.id, user) == false
```
Comment (hpihkala) [inline test/unit/com/unifina/service/EthereumIntegrationKeyServiceSpec.groovy]: throws

File: test/unit/com/unifina/service/StreamServiceSpec.groovy [hunk 32]
```diff
@@ -32,6 +33,14 @@ class StreamServiceSpec extends Specification {
 
 	SecUser me = new SecUser(username: "me")
 
+	// This gets the real services injected into the filters
+	// From https://github.com/grails/grails-core/issues/9191
+	static doWithSpring = {
+		springSecurityService(SpringSecurityService)
```
Comment (hpihkala) [inline test/unit/com/unifina/service/StreamServiceSpec.groovy]: repeat fix

File: test/functional/AccessControlCoreSpec.groovy [hunk 33]
```diff
@@ -30,7 +30,7 @@ class AccessControlCoreSpec extends LoginTester2Spec implements ConfirmationMixi
 	}
 
 	def "user cannot access admin pages"() {
-		when: "user logins"
+		when: "me logins"
```
Comment (hpihkala) [inline test/functional/AccessControlCoreSpec.groovy]: Uh-oh, looks like some global text replace user -> me has touched a lot of files in this commit?

File: grails-app/views/help/api/markdown/introduction.md [hunk 34]
```diff
@@ -17,12 +17,44 @@ When reading from or writing to Streams, you can use a Stream-specific anonymous
...
+    "challenge": {
+	    "id": "challenge-id",
+	    "challenge": "challenge-text-that-you-signed"
+    },
+    "signature": "signature-of-the-challenge",
+    "address": "your-public-ethereum-address"
+}
+```
+
+The signature must follow the convention described [here](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md). The secp256k1 ECDSA algorithm is applied on the keccak256 hash of a string derived from the challenge text:
```
Comment (hpihkala) [inline grails-app/views/help/api/markdown/introduction.md]: If the content to be signed will be `challenge + address`, then it needs to be mentioned here

File: grails-app/controllers/com/unifina/controller/api/LoginApiController.groovy [hunk 35]
```diff
@@ -53,7 +53,7 @@ class LoginApiController {
 
 	@StreamrApi(authenticationLevel = AuthLevel.NONE)
 	def apikey(ApiKeyCommand cmd) {
-		SecUser user = userService.getUserFromApiKey(cmd.apiKey)
+		SecUser user = userService.getUserFromApiKey(cmd.apikey)
```
Comment (hpihkala) [inline grails-app/controllers/com/unifina/controller/api/LoginApiController.groovy]: I just realized that in `streamr-client` we use `apiKey`, right? Would be good to be consistent, also all-lower-caps looks kinda funny. Would it be a nightmare to change it back? I apologize if I earlier gave the opposite advice!

File: grails-app/services/com/unifina/service/EthereumIntegrationKeyService.groovy [hunk 36]
```diff
@@ -115,30 +114,33 @@ class EthereumIntegrationKeyService {
...
+			accountLocked  : false,
+			passwordExpired: false
+		])
+		new IntegrationKey(
+			name: address,
+			user: user,
+			service: IntegrationKey.Service.ETHEREUM_ID,
+			idInService: address,
+			json: ([
+				address: new String(address)
```
[the remark this answers is not in the document]
Comment (mthambipillai) [reply grails-app/services/com/unifina/service/EthereumIntegrationKeyService.groovy]: I don't know I didn't write this code, I just moved it. But I can try to change it.

File: grails-app/services/com/unifina/service/ChallengeService.groovy [hunk 37]
```diff
@@ -30,34 +23,18 @@ class ChallengeService {
...
 	String verifyChallengeAndGetAddress(String challengeID, String challenge, String signature) {
-		String challengeRedis = getConnection().get(challengeID)
+		String challengeRedis = keyValueStoreService.get(challengeID)
 		boolean invalidChallenge = challengeRedis == null || challenge != challengeRedis
 		if (invalidChallenge) {
 			return null
-		} else {
-			getConnection().del(challengeID)
 		}
 		byte[] messageHash = ECRecover.calculateMessageHash(challenge)
```
[the remark this answers is not in the document]
Comment (mthambipillai) [reply grails-app/services/com/unifina/service/ChallengeService.groovy]: I was thinking of doing it in a separate branch after this one is merged because it is getting heavy. But if you prefer I can implement it in this branch.

File: grails-app/controllers/com/unifina/controller/api/LoginApiController.groovy (hunk 35 above)
Comment (hpihkala) [reply grails-app/controllers/com/unifina/controller/api/LoginApiController.groovy]: @mthambipillai Maybe we're looking at some different places, but for the config option you give to `StreamrClient` is `apiKey`. See e.g. [this line](https://github.com/streamr-dev/streamr-client/blob/master/src/StreamrClient.js#L25)?

File: grails-app/services/com/unifina/service/ChallengeService.groovy (hunk 37 above)
[the remark this answers is not in the document]
Comment (hpihkala) [reply grails-app/services/com/unifina/service/ChallengeService.groovy]: @mthambipillai It should be just a small change? Let's discuss when we meet today. I think we need to include it if that's what we're going for - otherwise we end up shipping this update and then making an incompatible change right after.

File: grails-app/services/com/unifina/service/ChallengeService.groovy [hunk 38]
```diff
@@ -30,34 +23,18 @@ class ChallengeService {
 		}
```
[the remark this answers is not in the document]
Comment (hpihkala) [reply grails-app/services/com/unifina/service/ChallengeService.groovy]: You would be able to remove the error rendering from `LoginApiController#response` if you change the semantics of this service to throw when unsuccessful (perhaps make it `void ensureValidResponse(...)`)

File: grails-app/services/com/unifina/service/UserService.groovy [hunk 39]
```diff
@@ -161,24 +163,24 @@ class UserService {
 		}
 	}
 
-	SecUser getUserFromUsernameAndPassword(String username, String password) {
+	SecUser getUserFromUsernameAndPassword(String username, String password) throws InvalidUsernameAndPasswordException {
 		PasswordEncoder encoder = new BCryptPasswordEncoder()
 		SecUser user = SecUser.findByUsername(username)
 		if (user == null) {
-			return null
+			throw new InvalidUsernameAndPasswordException("Invalid: "+username+", "+password)
```
Comment (hpihkala) [inline grails-app/services/com/unifina/service/UserService.groovy]: Let's not echo back the credentials! They could accidentally end up in our server logs for example. The message should be just vague standard message, such as _Invalid username or password._

File: grails-app/services/com/unifina/service/UserService.groovy [hunk 40]
```diff
@@ -161,24 +163,24 @@ class UserService {
...
 		}
 		String dbHash = user.password
 		if (encoder.matches(password, dbHash)) {
 			return user
 		}else {
-			return null
+			throw new InvalidUsernameAndPasswordException("Invalid: "+username+", "+password)
```
Comment (hpihkala) [inline grails-app/services/com/unifina/service/UserService.groovy]: here too

File: grails-app/services/com/unifina/service/ChallengeService.groovy (hunk 38 above)
[the remark this answers is not in the document]
Comment (mthambipillai) [reply grails-app/services/com/unifina/service/ChallengeService.groovy]: Yes I thought the same. I can do that if you want, but it didn't seem very elegant to have a function returning nothing and that persists nothing, just a function that throws or not. A function with minimum side effects returning a boolean for validity seemed more explicit to me, where as a `void ensureValidResponse()` needs to be inspected to understand what it does.

File: grails-app/services/com/unifina/service/ChallengeService.groovy [hunk 41]
```diff
@@ -7,36 +7,42 @@ import com.unifina.security.Challenge
...
-			boolean valid = address.toLowerCase() == publicKey.toLowerCase()
-			if (valid) {
-				keyValueStoreService.delete(challengeID)
-			}
-			return valid
-		} catch(ChallengeVerificationFailedException e) {
-			return false
+	void checkValidChallengeResponse(String challengeID, String challenge, String signature, String givenAddress) throws ChallengeVerificationFailedException {
+		String address = verifyChallengeAndGetAddress(challengeID, challenge, signature)
+		boolean valid = address.toLowerCase() == givenAddress.toLowerCase() && challengeTextContainsAddress(challenge, givenAddress)
```
[the remark this answers is not in the document]
Comment (mthambipillai) [reply grails-app/services/com/unifina/service/ChallengeService.groovy]: Yes it is needed to prove private key ownership. Otherwise the following scenario is possible:

- User generates random address `A` and makes a `POST` on `api/v1/login/challenge/A`.
- Backend generates challenge text `ch` that contains address `A` and send it back to user.
- User generates a random signature `S` and ECrecover `A'` from `ch` and `S`.
- User sends `ch`, `S` and `A'` as his own address to `api/v1/login/response`.
- Backend checks that the received challenge text `ch` is the same that the one stored. It is.
- Backend ECrecover an address `A''` from `ch` and `S` and checks that `A'' == A'`. They are equal so user gets logged in.

Basically we need to remember the address given in the first place. But now that I think about it, the user has to give his address twice. I could change the spec to drop the address in the `api/v1/login/response`, just ECrecover `A'` from `ch` and `S` and extract `A` from `ch`, then check `A == A'`.

[the remark this answers is not in the document]
Comment (hpihkala) [reply grails-app/services/com/unifina/service/ChallengeService.groovy]: @mthambipillai Sorry, my question was unclear. What I meant was, should we have:

`challengeTextContainsAddress(challenge, givenAddress.toLowerCase())`

instead of what you currently have (without `toLowerCase`):

`challengeTextContainsAddress(challenge, givenAddress)`

**Comment by hpihkala** [comment]
@mthambipillai LGTM, feel free to land this one if you're happy with it. Please remember to squash-and-merge.
